packages/astro/src/core/build/plugins/plugin-content.ts

```diff
@@ -0,0 +1,302 @@
+import { normalizePath, type Plugin as VitePlugin } from 'vite';
+import fsMod from 'node:fs';
+import { createHash } from 'node:crypto';
+import { addRollupInput } from '../add-rollup-input.js';
+import { type BuildInternals } from '../internal.js';
+import type { AstroBuildPlugin } from '../plugin.js';
+import type { StaticBuildOptions } from '../types.js';
+import { generateContentEntryFile, generateLookupMap } from '../../../content/vite-plugin-content-virtual-mod.js';
+import { joinPaths, removeFileExtension, removeLeadingForwardSlash } from '../../path.js';
+import { fileURLToPath } from 'node:url';
+import { type ContentLookupMap, hasContentFlag } from '../../../content/utils.js';
+import { CONTENT_RENDER_FLAG, PROPAGATED_ASSET_FLAG } from '../../../content/consts.js';
+import { copyFiles } from '../static-build.js';
+import pLimit from 'p-limit';
+import { extendManualChunks } from './util.js';
+import { isServerLikeOutput } from '../../../prerender/utils.js';
+import { encodeName } from '../util.js';
+
+const CONTENT_CACHE_DIR = './content/';
+const CONTENT_MANIFEST_FILE = './manifest.json';
+// IMPORTANT: Update this version when making significant changes to the manifest format.
+// Only manifests generated with the same version number can be compared.
+const CONTENT_MANIFEST_VERSION = 0;
+
+interface ContentManifestKey {
+	collection: string;
+	type: 'content' | 'data';
+	entry: string;
+}
+interface ContentManifest {
+	version: number;
+	entries: [ContentManifestKey, string][];
+	// Tracks components that should be included in the server build
+	// When the cache is restored, these might no longer be referenced
+	serverEntries: string[];
+	// Tracks components that should be passed to the client build
+	// When the cache is restored, these might no longer be referenced
+	clientEntries: string[];
+}
+
+const virtualEmptyModuleId = `virtual:empty-content`;
+const resolvedVirtualEmptyModuleId = `\0${virtualEmptyModuleId}`;
+
+function createContentManifest(): ContentManifest {
+	return { version: -1, entries: [], serverEntries: [], clientEntries: [] };
+}
+
+function vitePluginContent(opts: StaticBuildOptions, lookupMap: ContentLookupMap, internals: BuildInternals): VitePlugin {
+	const { config } = opts.settings;
+	const { cacheDir } = config;
+	const distRoot = config.outDir;
+	const distContentRoot = new URL('./content/', distRoot);
+	const cachedChunks = new URL('./chunks/', opts.settings.config.cacheDir);
+	const distChunks = new URL('./chunks/', opts.settings.config.outDir);
```

Comment:
nit: we extract `config`, but then we use `opts.settings.config` later. Or we use `distRoot` AND `opts.settings.config.outDir` which are the same. It can get confusing, so maybe we should try to use the same data everywhere